### DISCUSSION TITLE
What is the proper strategy for running this inside GitLab CI/CD?

### DISCUSSION THREAD

**Discussion by doulighan** [mtkennerly/poetry-dynamic-versioning#128] [Q&A] [answered]
Having trouble coming up with a good solution here.

When you push a tag to gitlab, it starts a tag pipeline. The tag pipeline a) has no associated branch, and b) uses a detached head. 

This means the `git for-each-ref --merged master` call made by poetry-dynamic-versioning fails. The detatched head doesn't know about master, and gets a `malformed object name master` error. I can manually `checkout master` and the versioning will work, but the tag pipeline might itself not actually be associated with the master branch. I could have tagged `development` instead, which would also trigger a tag pipeline. Because the tag pipeline doesn't have any concept of a branch, I don't *really* know what to checkout before I run my `poetry build`. 

Additionally, if I manually run `git for-each-ref "refs/tags/**" --format "%(refname)" --sort -creatordate --merged HEAD`, then I get the proper tag output. Is there a way to tell the library to use `--merged HEAD` instead of `--merged master`?

Another option would be to stop using tag pipelines all-together. Instead, I run a pipeline on a merge or push to master. In this case I can be confident about the current branch, and let dynamic-versioning calculate the tag based on previous tag history. But this has the issue of not letting me simply tag a commit to create a release. I have to make sure the tag I want has already been created in history before the actual code changes are merged, it seems backwards and probably will lead to having to run multiple pipelines to get a proper release. Merge once to get the changes, manually tag the merge commit, then re-run the merge pipeline to actually trigger the versioning on that new tag. Definitely feels wonky.

I have a feeling I'm just thinking about this wrong. Does anyone have experience with this?

**Comment by mtkennerly** [comment, marked as answer]
Hi! Thanks for the detailed question.

> When you push a tag to gitlab, it starts a tag pipeline. The tag pipeline a) has no associated branch, and b) uses a detached head.

I'm not as familiar with GitLab as I am with GitHub, but I think you could fix it with some combination of these GitLab settings:

* `GIT_DEPTH: 0` in the `variables` section
* `git checkout "$CI_COMMIT_REF_NAME"` in the `before_script` section

> Is there a way to tell the library to use `--merged HEAD` instead of `--merged master`?

It uses [`HEAD` by default](https://github.com/mtkennerly/dunamai/blob/2e2d722bcf9624cc0b6407ba29e4735eaf0f7451/dunamai/__init__.py#L1050-L1051), and you can customize it with the `tag-branch` option. Where are you seeing it default to `master`?

**Reply by doulighan** [reply]
Ah thank you! Absolutely my mistake, I was specifying "master" to be the default tag_branch without knowing it. Appreciate the help!
